upgrader/upgrader.go

```diff
@@ -0,0 +1,106 @@
+package upgrader
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/NethermindEth/juno/utils"
+)
+
+type Upgrader struct {
+	client         *http.Client
+	log            utils.SimpleLogger
+	apiURL         string
+	currentVersion *semver.Version
+	releasesURL    string
+	delay          time.Duration
+}
+
+func NewUpgrader(version *semver.Version, apiURL, releasesURL string, delay time.Duration, log utils.SimpleLogger) *Upgrader {
+	return &Upgrader{
+		currentVersion: version,
+		client:         &http.Client{},
+		log:            log,
+		apiURL:         apiURL,
+		releasesURL:    releasesURL,
+		delay:          delay,
+	}
+}
+
+func (u *Upgrader) WithClient(client *http.Client) *Upgrader {
+	u.client = client
+	return u
+}
+
+func (u *Upgrader) WithLog(log utils.SimpleLogger) *Upgrader {
+	u.log = log
+	return u
+}
+
+type Release struct {
+	Version    *semver.Version `json:"tag_name"`
+	Draft      bool            `json:"draft"`
+	PreRelease bool            `json:"prerelease"`
+}
+
+func (u *Upgrader) Run(ctx context.Context) error {
+	timer := time.NewTimer(time.Millisecond) // Don't wait the first time.
+	defer timer.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			var req *http.Request
+			req, err := http.NewRequestWithContext(ctx, "GET", u.apiURL, http.NoBody)
+			if err != nil {
+				u.log.Debugw("Failed to create new request with context")
+				continue
+			}
+
+			resp, err := u.client.Do(req)
+			if err != nil {
+				u.log.Debugw("Failed to fetch latest release", "err", err)
+				continue
+			} else if resp.StatusCode != http.StatusOK {
+				u.log.Debugw("Failed to fetch latest release", "status", resp.Status)
+				continue
+			}
+
+			latest := new(Release)
+			if err := json.NewDecoder(resp.Body).Decode(&latest); err == nil {
+				if needsUpdate(*u.currentVersion, *latest.Version) {
+					u.log.Warnw("New release is available.",
+						"currentVersion", u.currentVersion.String(),
+						"newVersion", latest.Version.String(),
+						"link", u.releasesURL)
+				} else {
+					u.log.Debugw("Application is up-to-date.")
+				}
+			} else {
+				u.log.Debugw("Failed to unmarshal latest release")
+			}
+
+			timer.Reset(u.delay)
+			resp.Body.Close()
+		}
+	}
+}
+
+func needsUpdate(currentVersion, latestVersion semver.Version) bool {
+	if currentVersion.Major() == latestVersion.Major() {
```

Comment:
why can't we just use https://pkg.go.dev/github.com/Masterminds/semver#Version.GreaterThan

Reply:
It doesn't work well when the version isn't exactly semver (as when building from `main`). 

Will add to test cases to make this clear, and add a comment.